app/models/user_role.rb

```diff
@@ -139,10 +139,9 @@ def discourse_user_group
   end
 
   def can_user_read?(user)
-    # A user can view a role if:
-    # 1. the role belongs to a non-hidden group, or
-    # 2. the user has read-access to that group.
-    !group.is_hidden || user&.has_permission?(:can_read_groups, group.id)
+    return true unless group.is_hidden # Roles of non-hidden groups are public
+    return false if user.nil? # Roles of hidden groups are visible only to a set of users based on permisssions.
+    user&.has_permission?(is_active? ? :can_read_groups_current : :can_read_groups_past, group.id)
```

Comment:
The `user&.has_permission?` null-safe `&` is unnecessary because you check `if user.nil?` in the line directly above

Reply:
Also, the line would feel less busy if you extract the ternary check to a separate line and store it in a `role_permission` variable. Then just check `user.has_permission?(role_permission, group.id)`

Reply:
Okay applied both changes, thanks.